sdk/core/core/test/cmocka/test_az_url_encode.c

```diff
@@ -51,16 +54,112 @@ static void test_url_encode(void** state)
   az_span const buffer = AZ_SPAN_FROM_BUFFER(buf);
   int32_t url_length = 0;
 
-  assert_true(az_succeeded(
-      _az_span_url_encode(buffer, AZ_SPAN_FROM_STR("https://vault.azure.net"), &url_length)));
+  {
+    // Empty
+    uint8_t buf0[0] = {};
+    az_span const buffer0 = AZ_SPAN_FROM_BUFFER(buf0);
+    url_length = 0xFF;
+    assert_true(az_succeeded(_az_span_url_encode(buffer0, { 0 }, &url_length)));
+    assert_true(az_span_is_content_equal(az_span_slice(buffer0, 0, url_length), { 0 }));
+    assert_int_equal(url_length, 0);
+  }
+  {
+    // Just enough to do no encoding
+    uint8_t buf5[5] = { 0 };
+    az_span const buffer5 = AZ_SPAN_FROM_BUFFER(buf5);
+
+    url_length = 0xFF;
+    assert_true(az_succeeded(_az_span_url_encode(buffer5, AZ_SPAN_FROM_STR("hello"), &url_length)));
+
+    assert_true(
+        az_span_is_content_equal(az_span_slice(bufer5, 0, url_length), AZ_SPAN_FROM_STR("hello")));
+
+    assert_int_equal(url_length, sizeof("hello"));
+
+    // Insufficient size
+    url_length = 0xFF;
+    assert_true(
+        _az_span_url_encode(az_span_slice(0, 2), AZ_SPAN_FROM_STR("/"), &url_length)
+        == AZ_ERROR_INSUFFICIENT_SPAN_SIZE);
+
+    assert_int_equal(url_length, 0);
+    assert_true(az_span_is_content_equal(
+        az_span_slice(buffer5, 0, sizeof(buf5)), AZ_SPAN_FROM_STR("%2Flo")));
+  }
+  {
+    // May be sufficient to perform encoding
+    uint8_t buf7[7] = { 0 };
+    az_span const buffer7 = AZ_SPAN_FROM_BUFFER(buf7);
+
+    url_length = 0xFF;
+    assert_true(az_succeeded(_az_span_url_encode(buffer7, AZ_SPAN_FROM_STR("he/lo"), &url_length)));
+
+    assert_true(az_span_is_content_equal(
+        az_span_slice(buffer7, 0, url_length), AZ_SPAN_FROM_STR("he%2Flo")));
+
+    assert_int_equal(url_length, sizeof("he%2Flo"));
+  }
+#ifdef AZ_NO_PRECONDITION_CHECKING
+  {
+    uint8_t buf5[5] = { '*', '*', '*', '*', '*' };
+    az_span const buffer5 = AZ_SPAN_FROM_BUFFER(buf5);
+
+    url_length = 0xFF;
+    assert_true(
+        _az_span_url_encode(buffer5, AZ_SPAN_FROM_STR("hello_world"), &url_length)
+        == AZ_ERROR_INSUFFICIENT_SPAN_SIZE);
+
+    assert_int_equal(url_length, 0);
+
+    url_length = 0xFF;
+    assert_true(az_span_is_content_equal(
+        az_span_slice(buffer5, 0, sizeof(buf5)), AZ_SPAN_FROM_STR("*****")));
```

Comment:
The slice isn't necessary here.
```suggestion
        buffer5, AZ_SPAN_FROM_STR("*****")));
```

Reply:
Done.